### PR TITLE
Add readBlock to objectForKey & writeBlock to setObject:forKey:

### PR DESCRIPTION
**Pull request by dimazen**
Generally this is a merge of https://github.com/tumblr/TMCache/pull/30/commits to PINCache. 
Read & Write blocks are quite useful when object isn't compatible with NSCoding or manual action needs to be performed

### CODE REVIEW

**Comment by jparise** [comment]
Thanks for the contribution!  Would you be willing to split this up into multiple pull requests (e.g. one that adds `nonnull`, etc. attributes to the current code, one that adds the new read/write blocks, etc.)?  That would make it easier to review the functional changes on their own.

**Comment by dimazen** [comment]
@jparise should I remove nullability attributes and then make a diff branch with them? 
What is the best way to do it?

**Comment by jparise** [comment]
Yes, that would be my preference.  @garrettmoon is the primary maintainer, though, so let's see what he says.

**Comment by garrettmoon** [comment]
Hi Dima,




Thanks so much for the contribution!




I agree with Jon that splitting this up into multiple pull requests will make it easier to review and merge in. If you're willing to do that, it would be *greatly* appreciated!

**Comment by dimazen** [comment]
I'll do my best tomorrow's morning ;)

**Comment by dimazen** [comment]
Sorry, @garrettmoon, I'm a bit busy atm, so waiting for the weekends

**Comment by garrettmoon** [comment]
Hey, no worries!

**Comment by dimazen** [comment]
Guys, having a really crazy pre-release week, sorry one more time

**Comment by dimazen** [comment]
@garrettmoon I reverted nullability attributes, check it!

**Comment by garrettmoon** [comment]
Thanks! I'll give this another look by EOD Friday.

File: PINCache/PINCache.h [hunk 1]
```diff
@@ -124,6 +140,18 @@ typedef void (^PINCacheObjectBlock)(PINCache *cache, NSString *key, id __nullabl
 - (void)setObject:(id <NSCoding>)object forKey:(NSString *)key block:(nullable PINCacheObjectBlock)block;
 
 /**
+ Stores an object in the cache for the specified key. This method returns immediately and executes the
+ passed block after the object has been stored, potentially in parallel with other blocks on the <queue>.
```
Comment (garrettmoon) [inline PINCache/PINCache.h]: should be <concurrentQueue>

File: PINCache/PINCache.m [hunk 2]
```diff
@@ -295,31 +320,55 @@ - (NSUInteger)diskByteCount
...
-    
+
     if (object) {
         // update the access time on disk
         [_diskCache fileURLForKey:key block:NULL];
     } else {
-        object = [_diskCache objectForKey:key];
+        PINDiskCacheReadBlock diskReadBlock = nil;
+        if (readBlock) {
+            diskReadBlock = ^id (PINDiskCache *cache, NSString *diskKey, NSURL *fileURL){
```
Comment (garrettmoon) [inline PINCache/PINCache.m]: nit, but diskKey should be key to be consistent elsewhere

File: PINCache/PINCache.m [hunk 3]
```diff
@@ -295,31 +320,55 @@ - (NSUInteger)diskByteCount
...
     if (!key || !object)
         return;
-    
+
     [_memoryCache setObject:object forKey:key];
-    [_diskCache setObject:object forKey:key];
+
+    PINDiskCacheWriteBlock diskWriteBlock = nil;
+    if (writeBlock) {
+        diskWriteBlock = ^BOOL (PINDiskCache *cache, NSString *diskKey, NSURL *fileURL, id diskObject){
```
Comment (garrettmoon) [inline PINCache/PINCache.m]: same here, this should be key, not diskKey to be consistent

File: PINCache/PINDiskCache.h [hunk 4]
```diff
@@ -19,7 +19,11 @@ typedef void (^PINDiskCacheBlock)(PINDiskCache *cache);
  A callback block which provides the cache, key and object as arguments
  */
 
-typedef void (^PINDiskCacheObjectBlock)(PINDiskCache *cache, NSString *key, id <NSCoding>  __nullable object, NSURL *fileURL);
+typedef void (^PINDiskCacheObjectBlock)(PINDiskCache *cache, NSString *key, id __nullable object, NSURL *fileURL);
```
Comment (garrettmoon) [inline PINCache/PINDiskCache.h]: Can you update the comment to indicate that if a write / read block is not used, then the id must conform to NSCoding?

File: tests/PINCache.xcodeproj/xcshareddata/xcschemes/PINCacheTests.xcscheme [hunk 5]
```diff
@@ -8,7 +8,7 @@
       <BuildActionEntries>
          <BuildActionEntry
             buildForTesting = "YES"
-            buildForRunning = "NO"
+            buildForRunning = "YES"
```
Comment (garrettmoon) [inline tests/PINCache.xcodeproj/xcshareddata/xcschemes/PINCacheTests.xcscheme]: Is this change necessary? What does it do?

File: tests/PINCache.xcodeproj/xcshareddata/xcschemes/iOS.xcscheme [hunk 6]
```diff
@@ -28,7 +28,26 @@
       shouldUseLaunchSchemeArgsEnv = "YES"
       buildConfiguration = "Debug">
       <Testables>
+         <TestableReference
```
Comment (garrettmoon) [inline tests/PINCache.xcodeproj/xcshareddata/xcschemes/iOS.xcscheme]: Are these changes necessary? What do they do?

**Comment by garrettmoon** [comment]
This looks good! Aside from the few nits above, there's one other thing I'd request: many of the changes are whitespace changes that seem unnecessary and conflict with the existing whitespace patterns. Do you mind cleaning this up?

File: PINCache/PINCache.h (hunk 1 above)
Comment (dimazen) [inline PINCache/PINCache.h]: Copy-paste got me :)

File: tests/PINCache.xcodeproj/xcshareddata/xcschemes/iOS.xcscheme (hunk 6 above)
Comment (dimazen) [inline tests/PINCache.xcodeproj/xcshareddata/xcschemes/iOS.xcscheme]: I've added ability to run tests from iOS scheme by clicking cmd+u. 
Needs to be removed.

**Comment by dimazen** [comment]
@garrettmoon 
Check out my fixes!

**Comment by dimazen** [comment]
@garrettmoon 
Aside this PR can you also check other branches, that I have, maybe some of them can be useful for master repo:
[feature/extension](https://github.com/dimazen/PINCache/tree/feature/extension) - saves key extension to disk, when you're passing key like "key.png". This one is useful, when you're using QLPreviewController - it can't identify object content-type without extension. 

[feature/object_at_path](https://github.com/dimazen/PINCache/tree/feature/object_at_path) - allows you to set object at path. This is useful, when object is relatively big, so instead of explicit read and write NSFileManager used to simply copy / move file.

[feature/trim_by_keys](https://github.com/dimazen/PINCache/tree/feature/trim_by_keys) - this one is quite specific. It allows you to trim a defined set of keys to some date (heavily used in my current project cleanup of unused key with specific criteria)

File: tests/PINCache.xcodeproj/xcshareddata/xcschemes/iOS.xcscheme [hunk 7]
```diff
@@ -29,6 +29,15 @@
...
       </Testables>
+      <MacroExpansion>
+         <BuildableReference
+            BuildableIdentifier = "primary"
+            BlueprintIdentifier = "6629FFEC1A66B512009C10BD"
+            BuildableName = "PINCache.framework"
+            BlueprintName = "iOS"
+            ReferencedContainer = "container:PINCache.xcodeproj">
+         </BuildableReference>
+      </MacroExpansion>
```
Comment (garrettmoon) [inline tests/PINCache.xcodeproj/xcshareddata/xcschemes/iOS.xcscheme]: Is this necessary? What does it do?

**Comment by garrettmoon** [comment]
@dimazen Nice! And if you feel that the features would be widely useful to others, feel free to submit pull requests. One more question and I think we're there!

File: tests/PINCache.xcodeproj/xcshareddata/xcschemes/iOS.xcscheme (hunk 7 above)
Comment (dimazen) [inline tests/PINCache.xcodeproj/xcshareddata/xcschemes/iOS.xcscheme]: Strange it gets commited. 
I was trying to run tests from iOS target by cmd+u, however this one should be reverted..

**Comment by garrettmoon** [comment]
I think I dropped the ball on this. The last question was, can you reformat your patch so it doesn't include all the whitespace changes? I know this is a pain, but it will help keep our git history. Thank you in advance!

**Comment by dimazen** [comment]
Argh, I've totally forgot about it, sure I will with no questions ;)

**Comment by garrettmoon** [comment]
Just to keep things clean, I'm going to close this. Feel free to reopen if you'd like to continue working on it.

**Comment by garrettmoon** [comment]
Or rather, let me know if you'd like to reopen :)
